### DISCUSSION TITLE
RangeQuery and WindowedRangeQuery in SQLAlchemy 2.0 syntax

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/sqlalchemy#7948] [Usage Questions]
Hello, I'm trying to adapt [WindowedRangeQuery](https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery) from the wiki to SQLAlchemy 2.0 syntax but I'm struggling on getting it to work. Is it possible to get a snippet of it using the new syntax?

**Comment by zzzeek** [comment]
Here's the first example on that page formatted as a diff, to get an idea of what changes (which is very little):

```diff
--- test4.py	2022-04-19 19:37:22.622960697 -0400
+++ test3.py	2022-04-19 19:41:28.748789595 -0400
@@ -1,29 +1,25 @@
-def windowed_query(q, column, windowsize):
+def windowed_query(s, q, column, windowsize):
     """"Break a Query into chunks on a given column."""
 
-    single_entity = q.is_single_entity
-    q = q.add_column(column).order_by(column)
+    q = q.add_columns(column).order_by(column)
     last_id = None
 
     while True:
         subq = q
         if last_id is not None:
             subq = subq.filter(column > last_id)
-        chunk = subq.limit(windowsize).all()
+        chunk = s.execute(subq.limit(windowsize)).all()
         if not chunk:
             break
         last_id = chunk[-1][-1]
         for row in chunk:
-            if single_entity:
-                yield row[0]
-            else:
-                yield row[0:-1]
+            yield row[0]
 
 
 if __name__ == "__main__":
-    from sqlalchemy import Column, Integer, create_engine
+    from sqlalchemy import Column, Integer, create_engine, select
     from sqlalchemy.orm import Session
-    from sqlalchemy.ext.declarative import declarative_base
+    from sqlalchemy.orm import declarative_base
     import random
 
     Base = declarative_base()
@@ -45,7 +41,7 @@
     s.add_all([Widget(id=i, data=j) for i, j in enumerate(data, 1)])
     s.commit()
 
-    q = s.query(Widget)
+    q = select(Widget)
 
-    for widget in windowed_query(q, Widget.data, 1000):
+    for widget in windowed_query(s, q, Widget.data, 1000):
         print("data:", widget.data)
\ No newline at end of file
```

**Reply by zzzeek** [reply]
the only change in functionality is that the above change will yield just the first entry of each row, which is the Widget object in that specific example.  options can be added to yield full rows also.

**Reply by michaelbukachi** [reply]
Thanks for this! :raised_hands:  I have a question though, I've noticed that passing in a query with an already defined sort affects the number of rows being yielded. :thinking: Have you faced this before?

**Reply by zzzeek** [reply]
I think that could interfere, sure, as you'd want the "windowed" ordering to be first in the list of ORDER BY.  Otherwise, other columns in your predefined ordering would take precedence and the windowed ordering would not produce a totally ordered set, meaning the windows would lose entries.   so....you need to work it out so that the windowed order by col is first in the list of ORDER BY.

**Reply by Denhai** [reply]
How about the second example on that page. How to refactor the `from_self` to `aliased`?

```py
q = session.query(
        column, 
        func.row_number().\
                over(order_by=column).\
                label('rownum')
        )

column_aliased = aliased(column, q.subquery())
# AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Widget.data has an attribute 'mapper'
```

**Reply by zzzeek** [reply]
oh is there a from_Self() on there?  that's not good

**Reply by zzzeek** [reply]
okey doke, the whole page is rewritten for 2.0, see https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery

**Reply by Denhai** [reply]
There seems to be an issue with [this fix](https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery/_compare/3539941e9cfc3451157291973864270c57579c8b):
`SAWarning: SELECT statement has a cartesian product between FROM element(s) "widget" and FROM element "anon_1".  Apply join condition(s) between each element to resolve.`

**Reply by CaselIT** [reply]
that seems to be the case. My fix was not really correct. Will take another look

**Reply by CaselIT** [reply]
I've fixed it in the last change https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery/_compare/79b8819f7129e880cedf4016166a40c5ece231e2?diff=split

@zzzeek can you give a check?

**Reply by zzzeek** [reply]
it looks much more complicated.   the way I had it, getting `.columns[-1]` should be guaranteed to work since it just added that column on the line above.

**Reply by CaselIT** [reply]
but you can't use rownum in the where with data. that makes no sense

the result of that query is used here:
```py
    intervals = list(session.scalars(new_stmt))

    # yield out WHERE clauses for each range
    while intervals:
        start = intervals.pop(0)
        if intervals:
            yield and_(column >= start, column < intervals[0])
        else:
            yield column >= start
```

you can't use row_num to do `column >= start`

**Reply by zzzeek** [reply]
OK the correct query is from the old version and looks like this:

```sql
SELECT anon_1.widget_data AS anon_1_widget_data 
FROM (SELECT widget.data AS widget_data, row_number() OVER (ORDER BY widget.data) AS rownum 
FROM widget) AS anon_1 
WHERE rownum %% 1000=1
```

so yes it should select the data, rownum is relative to just the window query.

I've made a change against my previous version.   The typing has an ignore right now due to #10326.   so it's just the one line plus the query comment.

this is the diff:  https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery/_compare/aa93aacffe167531d4125a9f7b0975ad78ee052b...f4e2ca46549fb014f0b757f24250018253a8a1b4?source=1#diff-47af3b9eb55a01638eab3a50926f597cdfb9a08687c12b6f9397030b86a8a849

**Reply by CaselIT** [reply]
Yes that was what I wanted to do too, but typing complained and that's what prompted me to open the other issue
